proof/refine/ARM/Schedule_R.thy

```diff
@@ -3986,6 +3986,236 @@ lemma refillPopHead_valid_refills'[wp]:
+                         readRefillSize_def refillIndex_def opt_map_red readSchedContext_def
+                  dest!: readObject_ko_at'_sc split: option.splits)
+  done
+
+lemma refill_head_overlapping_simp:
+  "sc_at sc_ptr s \<Longrightarrow> refill_head_overlapping sc_ptr s =
+     (scs_of2 s ||> (\<lambda>sc. Suc 0 < length (sc_refills sc)
+                         \<and> r_time (hd (tl (sc_refills sc)))
+                                   \<le> r_time (refill_hd sc) + r_amount (refill_hd sc))) sc_ptr"
+  unfolding refill_head_overlapping_def
+  apply (insert no_ofailD[OF no_ofail_read_sched_context])
+  apply (clarsimp simp: obind_def obj_at_def is_sc_obj opt_map_red
+                 split: option.split)
+  apply (drule_tac x=s and y=sc_ptr in meta_spec2)
+  apply (clarsimp dest!: read_sched_context_SomeD)
+  done
+
+lemma refillHeadOverlapping_corres_eq:
+  "\<lbrakk>(s, s') \<in> state_relation; sc_at sc_ptr s; sc_at' sc_ptr s'; valid_refills' sc_ptr s'\<rbrakk>
+   \<Longrightarrow> refill_head_overlapping sc_ptr s = refillHeadOverlapping sc_ptr s'"
+  apply (frule no_ofailD[OF no_ofail_refillHeadOverlapping])
+  apply (insert refillHeadOverlapping_implies_count_greater_than_one[of sc_ptr s'])
+  apply clarsimp
+  apply (drule (2) state_relation_sc_relation)
+  apply (clarsimp simp: obj_at_simps is_sc_obj)
+  apply (rename_tac b n sc sc')
+  apply (case_tac b;
+         clarsimp simp: refillHeadOverlapping_simp refill_head_overlapping_simp obj_at_simps
+                        is_sc_obj sc_relation_def valid_refills'_def refillHd_def
+                        neq_Nil_lengthI tl_drop_1 hd_drop_conv_nth refills_map_def hd_map
+                        hd_wrap_slice wrap_slice_index refill_map_def opt_map_red
+                 split: if_split_asm)
+  by linarith+
+
+lemma refillPopHead_scs_of':
+  "\<lbrace>\<lambda>s'. P (scs_of' s'(scp \<mapsto> (\<lambda>sc'. scRefillCount_update (\<lambda>_. scRefillCount sc' - Suc 0)
+                                      (scRefillHead_update
+                                          (\<lambda>_. refillNextIndex (scRefillHead sc') sc') sc'))
+                                        (the (scs_of' s' scp))))\<rbrace>
+  refillPopHead scp
+  \<lbrace>\<lambda>_ s'. P (scs_of' s')\<rbrace>"
+  unfolding refillPopHead_def
+  by (wpsimp wp: updateSchedContext_wp)
+
+crunches update_refill_hd, refill_pop_head, merge_refills
+  for is_active_sc2[wp]: "is_active_sc2 scp"
+  (wp: crunch_wps ignore: update_sched_context
+   simp: crunch_simps update_refill_hd_rewrite update_sched_context_set_refills_rewrite)
+
+lemma merge_refills_scs_of2[wp]:
+  "\<lbrace>\<lambda>s. P (scs_of2 s(scp \<mapsto> (\<lambda>sc. sc_refills_update
+            (\<lambda>_. merge_refill (refill_hd sc) (hd (tl (sc_refills sc))) # tl (tl (sc_refills sc))) sc)
+                          (the (scs_of2 s scp)))) \<rbrace>
+   merge_refills scp
+   \<lbrace>\<lambda>_ s. P (scs_of2 s)\<rbrace>"
```

Comment:
Out of curiosity, do you remember if there was a reason that this lemma was ok for `[wp]`, but the previous `refillPopHead_scs_of'` one wasn't?

Reply:
No apparent reason, the previous one is now added to the wpset.